rm-portal/src/main/java/org/ow2/proactive_grid_cloud_portal/rm/client/TreeView.java

```diff
@@ -46,73 +45,76 @@
 import com.smartgwt.client.widgets.tree.TreeGrid;
 import com.smartgwt.client.widgets.tree.TreeGridField;
 import com.smartgwt.client.widgets.tree.TreeNode;
-import com.smartgwt.client.widgets.tree.events.NodeClickEvent;
-import com.smartgwt.client.widgets.tree.events.NodeClickHandler;
 
 
 /**
  * Displays current nodes in a hierarchical tree view
  * <p>
  * NodeSource > Host > Node
- * 
- * 
- * 
- * 
- * @author mschnoor
  *
+ * @author mschnoor
  */
 public class TreeView implements NodesListener, NodeSelectedListener {
 
+    private static final String NODE_ID = "nodeId";
+
     private RMController controller = null;
 
-    /** tree view */
+    /**
+     * tree view
+     */
     private TreeGrid treeGrid = null;
 
-    /** tree data */
+    /**
+     * tree data
+     */
     private Tree tree = null;
 
-    /** parameter for {@link #nodesUpdated(Map)} last time it was called */
-    private Map<String, NodeSource> oldNodes = null;
-
-    /** treenodes currently held by {@link #tree} */
-    private HashMap<String, TreeNode> curNodes = null;
+    /**
+     * treenodes currently held by {@link #tree}
+     */
+    private HashMap<String, TreeNode> currentNodes = null;
 
-    /** prevent event cycling */
+    /**
+     * prevent event cycling
+     */
     private boolean ignoreNodeSelectedEvent = false;
 
     private class TNode extends TreeNode {
         Node rmNode = null;
 
-        public TNode(String name, Node node) {
+        TNode(String name, Node node) {
             super(name);
             this.rmNode = node;
+            this.setAttribute(NODE_ID, node.getNodeUrl());
         }
     }
 
     private class THost extends TreeNode {
         Host rmHost = null;
 
-        public THost(String name, Host h) {
+        THost(String name, Host h) {
             super(name);
             this.rmHost = h;
+            this.setAttribute(NODE_ID, h.getId());
         }
     }
 
     private class TNS extends TreeNode {
         NodeSource rmNS = null;
 
-        public TNS(String name, NodeSource ns) {
+        TNS(String name, NodeSource ns) {
             super(name);
             this.rmNS = ns;
+            this.setAttribute(NODE_ID, ns.getSourceName());
         }
     }
 
     TreeView(RMController controller) {
         this.controller = controller;
         this.controller.getEventDispatcher().addNodesListener(this);
         this.controller.getEventDispatcher().addNodeSelectedListener(this);
-        this.curNodes = new HashMap<String, TreeNode>();
-        this.oldNodes = new HashMap<String, NodeSource>();
+        this.currentNodes = new HashMap<>();
     }
 
     Canvas build() {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 22 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)